### PR TITLE
Add update sequence by sample and type end point 

### PR DESCRIPTION
**Pull request by vivbak**
FYI @violetbrina

### CODE REVIEW

File: api/routes/sequence.py [hunk 1]
```diff
@@ -134,7 +159,7 @@ async def get_latest_sequence_id_from_sample_id_and_type(
     '/all-for-sample-id/{sample_id}',
     operation_id='getAllSequencesForSampleId',
 )
-async def get_all_sequence_id_for_sample_id(
+async def get_all_sequences_id_for_sample_id(
```
Comment (vivbak) [inline api/routes/sequence.py:162]: @violetbrina I don't remember doing this. Was it on purpose? 😅

Comment (violetbrina) [reply api/routes/sequence.py]: yeah fixing a typo

**Review by violetbrina — APPROVED** [review]
Lookin good

File: api/routes/sequence.py [hunk 2]
```diff
@@ -65,6 +65,51 @@ async def update_sequence(
...
+        sequence_type=sequence_type,
+        status=sequence.status,
+        meta=sequence.meta,
+    )
+
+    return sequence_id
+
+
+@router.patch(
+    '/{external_id}/{sequence_type}', operation_id='updateSequenceFromSampleAndType'
```
Comment (illusional) [inline api/routes/sequence.py:92]: These two router URLs are effectively the same, and fastapi won't know which one to use, ie: `{placeholder}/{placeholder}`. Could you add some thing to make it more unique, eg: `/sample/internal/{internal_id}/{seqtype}` `/sample/external/{external_id}/{seqtype}`. Or use the same method name, but add it as a query param:

`/sample/seqtype` and add params for `cpg_id` / `external_sample_id`.

File: test/test_sequence.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        # Create new sequence
+        _ = await self.seql.insert_sequencing(
+            self.sample_id,
+            SequenceType(self.sequence_type),
+            SequenceStatus(self.sequence_status),
+            {},
+        )
+
+    @run_test_as_sync
+    async def test_update_sequence_from_sample_and_type(self):
```
Comment (illusional) [inline test/test_sequence.py:48]: Could you add a test for no sample present, and sample present but no sequence available?

Comment (illusional) [inline test/test_sequence.py:48]: Could you add a test for no sample present, and sample present but no sequence available?

File: api/routes/sequence.py (hunk 2 above)
Comment (violetbrina) [reply api/routes/sequence.py:92]: Ah great point, Yeah I can rename one of them

File: test/test_sequence.py (hunk 3 above)
Comment (vivbak) [reply test/test_sequence.py:60]: Done!

Comment (vivbak) [reply test/test_sequence.py:60]: Done!

**Review by illusional — CHANGES_REQUESTED** [review]

File: test/test_sequence.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+
+        # Create new sample
+        await self.sl.insert_sample(
+            self.external_sample_id,
+            SampleType('blood'),
+            active=True,
+            meta={'Testing': 'test_sequence'},
+        )
+
+        sample_ids = await self.sl.get_sample_id_map_by_external_ids(
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply test/test_sequence.py:38]: Sure thing

File: test/test_sequence.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+            [self.external_sample_id], project=None
+        )
+        self.sample_id = sample_ids[self.external_sample_id]
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply test/test_sequence.py:41]: Done!

File: api/routes/sequence.py [hunk 6]
```diff
@@ -65,6 +65,53 @@ async def update_sequence(
...
+    sample_id: str,
+    sequence_type: SequenceType,
+    sequence: SequenceUpdateModel,
+    connection: Connection = get_projectless_db_connection,
+):
+    """Update the latest sequence by sample_id and sequence type"""
+    sequence_layer = SampleSequenceLayer(connection)
+    sample_id_raw = sample_id_transform_to_raw(sample_id)
+
+    sequence_id = await sequence_layer.update_sequence_from_sample_and_type(
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply api/routes/sequence.py:82]: Already built-in given that we make a call to `get_latest_sequence_id_from_sample_id_and_type` which does all of the project checks before calling the update in the table sequence.py

File: api/routes/sequence.py [hunk 7]
```diff
@@ -65,6 +65,53 @@ async def update_sequence(
...
+@router.patch(
+    '/external_id/{external_id}/{sequence_type}',
+    operation_id='updateSequenceFromSampleAndType',
+)
+async def update_sequence_from_external_id_and_type(
+    external_id: str,
+    sequence_type: SequenceType,
+    sequence: SequenceUpdateModel,
+    connection: Connection = get_projectless_db_connection,
+):
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply api/routes/sequence.py:101]: Done

File: db/python/layers/sequence.py [hunk 8]
```diff
@@ -295,3 +296,36 @@ async def upsert_sequences(self, iid: int, sequences: List[SequenceUpsert]):
...
+
+        return sequence_id
+
+    async def update_sequence_from_external_id_and_type(
+        self, external_id, sequence_type, status, meta
+    ):
+        """Update a sequence from the external_id and sequence_type"""
+
+        # Convert the external_id to an internal sample_id
+        sample_ids = await self.sampt.get_sample_id_map_by_external_ids(
```
[the remark this answers is not in the document]
Comment (violetbrina) [reply db/python/layers/sequence.py:324]: Changed

**Review by illusional — APPROVED** [review]

**Pull request merged**